src/IO/WriteBufferFromS3.cpp

```diff
@@ -83,7 +93,60 @@ void WriteBufferFromS3::nextImpl()
     if (temporary_buffer->tellp() == -1)
         allocateBuffer();
 
-    temporary_buffer->write(working_buffer.begin(), offset());
+    size_t size = offset();
+    temporary_buffer->write(working_buffer.begin(), size);
+
+    ThreadGroupStatusPtr running_group = CurrentThread::isInitialized() && CurrentThread::get().getThreadGroup()
+            ? CurrentThread::get().getThreadGroup()
+            : MainThreadStatus::getInstance().getThreadGroup();
+
+    if (CurrentThread::isInitialized())
+        query_context = CurrentThread::get().getQueryContext();
+
+    if (!query_context)
+    {
+        if (!shared_query_context)
+        {
+            ContextPtr global_context = CurrentThread::isInitialized() ? CurrentThread::get().getGlobalContext() : nullptr;
+            if (global_context)
+            {
+                shared_query_context = Context::createCopy(global_context);
+                shared_query_context->makeQueryContext();
+            }
+        }
+
+        if (shared_query_context)
+        {
+            shared_query_context->setCurrentQueryId(toString(UUIDHelpers::generateV4()));
+            query_context = shared_query_context;
+        }
+    }
+
+    if (cacheEnabled())
+    {
+        if (blob_name.empty())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Empty blob name");
+
+        auto cache_key = cache->hash(blob_name);
+        file_segments_holder.emplace(cache->setDownloading(cache_key, current_download_offset, size));
+        current_download_offset += size;
+
+        size_t remaining_size = size;
+        for (const auto & file_segment : file_segments_holder->file_segments)
+        {
+            size_t current_size = std::min(file_segment->range().size(), remaining_size);
+            remaining_size -= current_size;
+
+            if (file_segment->reserve(current_size))
+            {
+                file_segment->writeInMemory(working_buffer.begin(), current_size);
+            }
+            else
+            {
+                break;
```

Comment:
Need to delete file segments for which space reservation failed from FileSegmetnsHolder.